### ISSUE TITLE
savetxt complex number bug

### ISSUE THREAD

**Issue by drewvian** [numpy/numpy#7895]
I am using numpy version 1.11.0 in Python 2.7. I am working with numpy.complex number arrays file input and output. I have been having issues correctly saving and loading the arrays using the savetxt and loadtxt functions. When running the loadtxt function I received the following error:

``` python
/usr/lib/python2.7/dist-packages/numpy/lib/npyio.pyc in loadtxt(fname, dtype, comments, delimiter, converters, skiprows, usecols, unpack, ndmin)
    928 
    929             # Convert each value according to its column and store
--> 930             items = [conv(val) for (conv, val) in zip(converters, vals)]
    931             # Then pack it according to the dtype's nesting
    932             items = pack_items(items, packing)

/usr/lib/python2.7/dist-packages/numpy/lib/npyio.pyc in <lambda>(x)
    673         return floatconv
    674     elif issubclass(typ, np.complex):
--> 675         return lambda x: complex(asstr(x))
    676     elif issubclass(typ, np.bytes_):
    677         return bytes

ValueError: complex() arg is a malformed string
```

After much painful testing and searching, I located the problem. There is a problem with the negative imaginary values, where the savetxt function with the default format saves the value as:

``` python
 (6.800888281993669134e-05+-5.632350319387234187e-06j)
```

This makes the file unreadable using loadtext due to the '+-' invalid value.

The default format for complex values should be changed to something that allows proper parsing with loadtxt.

**Comment by simongibbons** [comment]
Agreed that would make sense, just getting rid of the extraneous `+` would be the most sensible thing to me.

I guess we need a deprecation cycle with this as it may break other people's code who rely on the precise format of the current `savetxt` output?

**Comment by charris** [comment]
I think this is a bug and that we could dispense with a deprecation. I can't offhand think of anything that expects `+-`. @simongibbons PR welcome.

**Comment by charris** [comment]
Does `savetxt` work for `long double`? The 16 digits might be too few, would like at least `repr` precision.

**Comment by SteveDoyle2** [comment]
I am running into this issue was well.  This is not a precision issue.  The `+` sign needs to be dynamic.

When I run the following code:

    import numpy as np
    A = np.array(
        [[1+2j, 1-2j],
         [2+3j, 2-3j]], dtype='complex128')
    
    np.savetxt('A.txt', A)
    Aprime = np.loadtxt('A.txt')
    assert np.array_equal(A, Aprime), 'A=\n%s\nAprime=\n%s' % (A, Aprime)

I get the following error

    Traceback (most recent call last):
      File "numpy_bug.py", line 7, in <module>
        Aprime = np.loadtxt('A.txt')
      File "F:\Anaconda\lib\site-packages\numpy\lib\npyio.py", line 930, in loadtxt
        items = [conv(val) for (conv, val) in zip(converters, vals)]
      File "F:\Anaconda\lib\site-packages\numpy\lib\npyio.py", line 659, in floatconv
        return float(x)
    ValueError: could not convert string to float:         
    (1.000000000000000000e+00+2.000000000000000000e+00j)

A.txt has the following data:

     (1.000000000000000000e+00+2.000000000000000000e+00j)  (1.000000000000000000e+00+-2.000000000000000000e+00j)
     (2.000000000000000000e+00+3.000000000000000000e+00j)  (2.000000000000000000e+00+-3.000000000000000000e+00j)

The `1-2j` and `2-3j` are written incorrectly as `1+-2j` and `2+-3j` respectively.  This is due to the following block that does not support negative imaginary parts:

    if iscomplex_X:
        fmt = [' (%s+%sj)' % (fmt, fmt), ] * ncol
    else:
        fmt = [fmt, ] * ncol

https://github.com/numpy/numpy/blob/v1.11.0/numpy/lib/npyio.py#L969-L1168

I do not see obvious way to fix this problem in `savetxt`, especially given that a user can provide a custom formatter.  I do see a fix in `loadtxt` though.

## old
    elif issubclass(typ, np.complex):
        return lambda x: complex(asstr(x))

## new
    elif issubclass(typ, np.complex):
        return lambda x: complex(asstr(x).replace('+-', '-'))

https://github.com/numpy/numpy/blob/v1.11.0/numpy/lib/npyio.py#L682-L966

Note that this issue is an extension of the following ticket.  The general case is not handled.

https://github.com/numpy/numpy/issues/1573

**Comment by valandil** [comment]
Our research group also encountered this issue. While it is easy to work around, it is extremely annoying that `np.savetxt` does not output data that `np.loadtxt` can understand. Could the suggestion by @SteveDoyle2 be implemented?

**Comment by WarrenWeckesser** [comment]
Another occurrence of the problem: https://stackoverflow.com/questions/47084008/numpy-genfromtxt-give-nan-for-complex-numbers-with-minus-signs

**Comment by mhauru** [comment]
+1: I, too, just spent an hour figuring out why a simple savetxt -> loadtxt/genfromtxt was failing, and discovered what @SteveDoyle2 described above.

**Comment by charris** [comment]
@ahaldane I believe you are fixing this?

**Comment by ahaldane** [comment]
It should be fixed, for sure. But I think it might need a rewrite of savetxt, and I would like to get some other things off my plate before trying to fix it. Please consider it up for grabs!

(I think the float128 problem from #10436 could be fixed at the same time, but again, I think it needs some kind of rewrite/new interface for savetxt, and it would be ambitious to aim for 1.14.1)

**Comment by charris** [comment]
OK, pushed off to 1.15, maybe later :)

**Comment by mattip** [comment]
Fixed by PR #10875, note that #10436 is still open
